pisa/stages/xsec/dis_sys.py

```diff
@@ -0,0 +1,213 @@
+        data=None,
+        params=None,
+        input_names=None,
+        output_names=None,
+        debug_mode=None,
+        input_specs=None,
+        calc_specs=None,
+        output_specs=None,
+        extrapolation_type='constant',
+    ):
+        expected_params = (
+            'dis_csms',
+        )
+        input_names = ()
+        output_names = ()
+
+        # what are the keys used from the inputs during apply
+        input_apply_keys = (
+            'dis_correction_total',
+            'dis_correction_diff',
+        )
+        # what keys are added or altered for the outputs during apply
+        output_apply_keys = (
+            'weights',
+        )
+
+        # init base class
+        super().__init__(
+            data=data,
+            params=params,
+            expected_params=expected_params,
+            input_names=input_names,
+            output_names=output_names,
+            debug_mode=debug_mode,
+            input_specs=input_specs,
+            calc_specs=calc_specs,
+            output_specs=output_specs,
+            input_apply_keys=input_apply_keys,
+            output_apply_keys=output_apply_keys,
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is None
+        assert self.output_mode is not None
+
+        self.extrapolation_type = extrapolation_type
+
+    @profile
+    def setup_function(self):
+
+        extrap_dict = from_file('cross_sections/tot_xsec_corr_Q2min1_isoscalar.pckl')
+
+        # load splines
+        wf_nucc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_CC_flat.pckl')
+        wf_nubarcc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_CC_flat.pckl')
+        wf_nunc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_NC_flat.pckl')
+        wf_nubarnc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_NC_flat.pckl')
+
+        # set this to events mode, as we need the per-event info to calculate these weights
+        self.data.data_specs = 'events'
+
+        for container in self.data:
+
+            # creat keys for external dict
+
+            if container.name.endswith('_cc'):
+                current = 'CC'
+            elif container.name.endswith('_nc'):
+                current = 'NC'
+            else:
+                raise ValueError('Can not determine whether container with name "%s" is pf type CC or NC based on its name'%container.name)
+            nu = 'Nu' if container['nubar'] > 0 else 'NuBar'
+
+            lgE = np.log10(container['true_energy'].get('host'))
+            bjorken_y = container['bjorken_y'].get('host')
+            dis = container['dis'].get('host')
+
+            #
+            # Calculate variation of total cross section
+            #
+
+            lgE_min = 2. if self.extrapolation_type == 'constant' else 1.68
+            mask = lgE <= lgE_min
+
+            w_tot = np.ones_like(lgE)
+            
+            poly_coef = extrap_dict[nu][current]['poly_coef']
```

Comment:
I think this block of code could be made more clear in what it's doing, more concise, and faster at the same time:

```python
poly_coef = extrap_dict[nu][current]['poly_coef']
lin_coef = extrap_dict[nu][current]['lin_coef']

if self.extrapolation_type == 'higher':
    w_tot = np.polyval(poly_coef, lgE)
else:
    poly_valid_mask = lgE >= lgE_min
    poly_invalid_mask = ~poly_valid_mask

    w_tot[poly_valid_mask] = np.polyval(poly_coef, lgE[poly_valid_mask])

    if self.extrapolation_type == 'constant':
        w_tot[poly_invalid_mask] = np.polyval(poly_coef, lgE_min)  # note Numpy broadcasts
    elif self.extrapolation_type == 'linear':
        w_tot[poly_invalid_mask] = np.polyval(lin_coef, lgE[poly_invalid_mask)
    else:
        raise ValueError('Unknown extrapolation type "%s"'%self.extrapolation_type)
```

Reply:
done